### DISCUSSION TITLE
Does Helidon MP support non-blocking REST calls with backpressure?

### DISCUSSION THREAD

**Discussion by khovanskiy** [helidon-io/helidon#3622] [Q&A] [answered]
According to the specification, we can use `AsyncRequest` and `@Suspended`, but all examples use an additional thread pool. So how can we set up the backpressure of incoming requests at the transport level (Netty)?
```
@Path("/example")
public class ExampleService {

    private static ExecutorService executor = Executors.newFixedThreadPool(10);

    @Path("/test")
    @GET
    public void test(
        @QueryParam("param1") String param1, 
        @QueryParam("param2") String param2, 
        @Suspended AsyncResponse ar
    ) {
        executor.execute( () -> {
            final var result = doSomeLongTask();
            ar.resume(result);
        });
    }
}
```

**Comment by spericas** [comment, marked as answer]
There is not support for that in Helidon MP. You should look at Helidon SE instead.

**Reply by khovanskiy** [reply]
Will this feature never be in Helidon MP?

**Reply by dalexandrov** [reply]
There is a way you to use Helidon SE service in MP:
https://helidon.io/docs/v2/#/mp/jaxrs/10_reactive-routing

**Reply by spericas** [reply]
Helidon MP is fundamentally a blocking API, even if you use `JerseySupport` with SE for a flavor of MP, it will not change that.

**Comment by tomas-langer** [comment]
1. Since Helidon 4.0, there is no Netty, no non-blocking event loop based implementation, and no need to use reactive programming (thanks to Virtual Thread feature in Java); backpressure is "natural", because we do not read data/write data unless somebody asks for it, or we can actually write it out (i.e. on incoming side, we only read from a socket if user action requests a read, on outgoing side, you code will block until we can write the data to the socket)
2. And since Helidon 4.0, you should not really used async responses, as any operation can be done with a proper `join()`, or `Future.get()` - even if you need to do some parallel task over you own executor service, just wait until it finishes, and then return the response; blocking is OK
3. In older versions, which were Netty based, the whole of MP is based on an ExecutorService (so it can be blocking), as reactive streams explicitly forbid any blocking operation; as a result, even reactive-like APIs of JAX-RS will not be "connected" to the Netty layer directly
